detach/detach.go

```diff
@@ -0,0 +1,106 @@
+package detach
+
+import (
+	"context"
+	"sync"
+	"time"
+)
+
+// Hooks are functions that are called with specific detach helpers to attach
+// values from the parent context to the new context. In order to be registered,
+//at least one function in Hooks must be non-nil. Nil functions will be skipped.
+type Hooks struct {
+	// Inline functions are used to attach values from the src context to the dst
+	// context in calls to Inline.
+	//
+	// Inline functions should assume that the calling function will wait for the
+	// any calls after Inline to return before it returns, so they may just
+	// copy values from src to dst.
+	Inline func(dst, src context.Context) context.Context
+
+	// Async functions wrap the run methods passed to Async. They should derive
+	// values from the src context to add to the dst context and do any
+	// setup/cleanup around those values.
+	//
+	// Async functions should assume that the calling function could return before
+	// they finish, so if context values on src could be closed or otherwise
+	// cleaned up, the wrapper should likely create new values and clean those up
+	// rather than just copying the values from src to dst.
+	Async func(dst, src context.Context, next func(ctx context.Context))
+}
+
+type hooksRegistry struct {
+	mu    sync.RWMutex
+	hooks []Hooks
+}
+
+func (r *hooksRegistry) add(hooks Hooks) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.hooks = append(r.hooks, hooks)
+}
+
+func (r *hooksRegistry) allHooks() []Hooks {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	return r.hooks
+}
```

Comment:
wait I don't think returning `r.hooks` and release the lock makes it safe to be used concurrently? to make it truly thread-safe we likely need to make a shallow copy here, e.g.:
```go
hooks := make([]Hooks, len(r.hooks))
copy(hooks, r.hooks)
return hooks
```

but in reality we expect all the `add` calls happen on initialization so this is not that big a deal.

Reply:
I could have sworn this already came up somewhere and I was told not to worry about copying it but I can't find the comment....... 🤔

Reply:
ah, @kylelemons and I talked about this outside of the PR,  his point was that since we are just appending, it would be safe to just return it. Thinking about it more, it _seems_ like a good idea to me to not leave that potential footgun around though?